### ISSUE TITLE
Weird ERROR Encountered When Running A Modified Taxi Service Tutorial Using "personFlow"

### ISSUE THREAD

**Issue by apecia** [eclipse-sumo/sumo#14104]
I modified the Taxi Service tutorial to use a random person distribution with "personFlows" and implement a simple taxi dispatch algorithm to meet the demand/supply while manipulating the "probability" value. When running different probability values, I've encountered two weird error that either quit the simulation or crash it. These errors happened randomly and out of nowhere which makes debugging my code very hard. To make sure that these errors happen randomly, I've run 60 different simulations with different parameters.

These are the two sample errors that I've encountered:
**1st Error:**
```
Warning: Teleporting person 'person.2' from vehicle destination edge 'gneE0' to intended destination edge 'gneE2' time=268.00
Quitting (on error)
```

**2nd Error:**
```
traci.exceptions.TraCIException: Could not interpret reservations for vehicle 'taxiV0' (Reservation id '0' is not known).
```

The following are the simulations I've run with different parameters:
Results running with 20s:
1. begin="0.00", end="20.00", probability="0.05" - 3 person loaded. Finished with no issue
2. begin="0.00", end="20.00", probability="0.1" - 5 person loaded. Finished with no issue
3. begin="0.00", end="20.00", probability="0.15" - 5 person loaded. Finished with no issue
4. begin="0.00", end="20.00", probability="0.2" - 6 person loaded. Finished with no issue
5. begin="0.00", end="20.00", probability="0.25" - 6 person loaded. Finished with no issue
6. begin="0.00", end="20.00", probability="0.3" - 7 person loaded. Finished with no issue
7. begin="0.00", end="20.00", probability="0.35" - 7 person loaded. Finished with no issue
8. begin="0.00", end="20.00", probability="0.4" - 9 person loaded. Finished with no issue
9. begin="0.00", end="20.00", probability="0.45" - 9 person loaded. Finished with no issue
10. begin="0.00", end="20.00", probability="0.5" - 10 person loaded. Finished with no issue
11. begin="0.00", end="20.00", probability="0.55" - 12 person loaded. Finished with no issue
12. begin="0.00", end="20.00", probability="0.6" - 16 person loaded. Quit with 1st error on steps 1575.
13. begin="0.00", end="20.00", probability="0.65" - 23 person loaded. Finished with no issue
14. begin="0.00", end="20.00", probability="0.7" - 29 person loaded. Quit with 1st error on steps 225.
15. begin="0.00", end="20.00", probability="0.75" - 23 person loaded. Quit with 2nd error on steps 375.
16. begin="0.00", end="20.00", probability="0.8" - 23 person loaded. Quit with 2nd error on steps 375.
17. begin="0.00", end="20.00", probability="0.85" - 24 person loaded. Quit with 2nd error on steps 425.
18. begin="0.00", end="20.00", probability="0.9" - 28 person loaded. Quit with 2nd error on steps 425.
19. begin="0.00", end="20.00", probability="0.95" - 27 person loaded. Quit with 2nd error on steps 575.
20. begin="0.00", end="20.00", probability="1.0" - 33 person loaded. Quit with 2nd error on steps 475.

Results running with 40s:
1. begin="0.00", end="40.00", probability="0.05" - 5 person loaded. Finished with no issue
2. begin="0.00", end="40.00", probability="0.1" - 7 person loaded. Finished with no issue
3. begin="0.00", end="40.00", probability="0.15" - 8 person loaded. Finished with no issue
4. begin="0.00", end="40.00", probability="0.2" - 21 person loaded. Quit with 2nd error on steps 375.
5. begin="0.00", end="40.00", probability="0.25" - 26 person loaded. Quit with 2nd error on steps 375.
6. begin="0.00", end="40.00", probability="0.3" - 24 person loaded. Quit with 2nd error on steps 375.
7. begin="0.00", end="40.00", probability="0.35" - 27 person loaded. Quit with 2nd error on steps 375.
8. begin="0.00", end="40.00", probability="0.4" - 30 person loaded. Quit with 2nd error on steps 475.
9. begin="0.00", end="40.00", probability="0.45" - 42 person loaded. Quit with 2nd error on steps 625.
10. begin="0.00", end="40.00", probability="0.5" - 41 person loaded. Quit with 1st error on steps 275.
11. begin="0.00", end="40.00", probability="0.55" - 43 person loaded. Quit with 1st error on steps 375.
12. begin="0.00", end="40.00", probability="0.6" - 45 person loaded. Quit with 1st error on steps 475.
13. begin="0.00", end="40.00", probability="0.65" - 51 person loaded. Quit with 2nd error on steps 475.
14. begin="0.00", end="40.00", probability="0.7" - 56 person loaded. Finished with no issue
15. begin="0.00", end="40.00", probability="0.75" - Crashed with 1st error at steps 1725.
16. begin="0.00", end="40.00", probability="0.8" - Crashed with 1st error at steps 1725.
17. begin="0.00", end="40.00", probability="0.85" - 64 person loaded. Quit with 1st error on steps 225.
18. begin="0.00", end="40.00", probability="0.9" - Crashed with 1st error at steps 375.
19. begin="0.00", end="40.00", probability="0.95" - Crashed with 1st error at steps 225.
20. begin="0.00", end="40.00", probability="1.0" - Crashed with 1st error steps 225.

Results running with 60s:
1. begin="0.00", end="60.00", probability="0.05" - 6 person loaded. Finished with no issue
2. begin="0.00", end="60.00", probability="0.1" - 16 person loaded. Quit with 2nd error on steps 475.
3. begin="0.00", end="60.00", probability="0.15" - 18 person loaded. Quit with 2nd error on steps 375.
4. begin="0.00", end="60.00", probability="0.2" - 22 person loaded. Quit with 2nd error on steps 425.
5. begin="0.00", end="60.00", probability="0.25" - 25 person loaded. Quit with 2nd error on steps 425.
6. begin="0.00", end="60.00", probability="0.3" - 29 person loaded. Finished with no issue
7. begin="0.00", end="60.00", probability="0.35" - 33 person loaded. Finished with no issue
8. begin="0.00", end="60.00", probability="0.4" - Crashed with 1st error at steps 1625.
9. begin="0.00", end="60.00", probability="0.45" - 47 person loaded. Quit with 2nd error on steps 475.
10. begin="0.00", end="60.00", probability="0.5" - 61 person loaded. Quit with 1st error on steps 275.
11. begin="0.00", end="60.00", probability="0.55" - 59 person loaded. Quit with 1st and 2nd error on steps 475.*
12. begin="0.00", end="60.00", probability="0.6" - 69 person loaded. Quit with 1st error on steps 425.
13. begin="0.00", end="60.00", probability="0.65" - 85 person loaded. Quit with 2nd error on steps 225.
14. begin="0.00", end="60.00", probability="0.7" - Crashed with 1st error at steps 325.
15. begin="0.00", end="60.00", probability="0.75" - 86 person loaded. Quit with 2nd error on steps 225.
16. begin="0.00", end="60.00", probability="0.8" - Crashed with 1st error at steps 275.
17. begin="0.00", end="60.00", probability="0.85" - 86 person loaded. Quit with 2nd error on steps 375.
18. begin="0.00", end="60.00", probability="0.9" - 104 person loaded. Quit with 1st error on steps 225.
19. begin="0.00", end="60.00", probability="0.95" - 103 person loaded. Quit with 1st and 2nd error on steps 375.*
20. begin="0.00", end="60.00", probability="1.0" - 113 person loaded. Quit with 1st and 2nd error on steps 475.*

The `add.add.xml` and `net.net.xml` are the same as the ones in the Taxi Service Tutorial. I only modified the `demand.rou.xml` to include `personFlow` and change some logics in the `runner.py`.

**The modified `demand.rou.xml`:**
````
 <!-- Route -->
<route edges="-gneE2 gneE0 -gneE0 gneE2 -gneE2" color="yellow" id="route_0"/>

<!-- Taxi-Type -->
<vType id="taxi" vClass="taxi" personCapacity="8">
    <param key="has.taxi.device" value="true"/>
</vType>

<!-- Person-Type -->
<vType id="pedestrian" vClass="pedestrian" width="0.6" length="0.263"/>

<!-- Other vehicles driving on streets -->
<vehicle id="vehicle_0" depart="0.00" route="route_0" color="blue"/>

<vehicle id="vehicle_1" depart="0.00" route="route_0" color="red"/>

<personFlow id="person" type="pedestrian" begin="0.00" end="60.00" color="green" probability="1.0">
    <walk from="-gneE1" busStop="busStop_gneE0_0_0"/>
    <ride from="gneE0" to="gneE2" lines="taxi"/>
</personFlow>
<!-- Other vehicles driving on streets -->
<vehicle id="vehicle_3" depart="130.00" route="route_0" color="#800080"/>
<vehicle id="vehicle_4" depart="134.00" route="route_0" color="#51414F"/>
````

**This is the modified `runner.py`:**
```
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Eclipse SUMO, Simulation of Urban MObility; see https://eclipse.dev/sumo
# Copyright (C) 2008-2023 German Aerospace Center (DLR) and others.
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# https://www.eclipse.org/legal/epl-2.0/
# This Source Code may also be made available under the following Secondary
# Licenses when the conditions for such availability set forth in the Eclipse
# Public License 2.0 are satisfied: GNU General Public License, version 2
# or later which is available at
# https://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
# SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later

# @file    runner.py
# @author  Michael Behrisch
# @author  Daniel Krajzewicz
# @author  Melanie Weber
# @author  Lara Kaiser
# @date    2021-03-29


from __future__ import absolute_import
from __future__ import print_function
import os
import sys
import optparse


# we need to import python modules from the $SUMO_HOME/tools directory
if 'SUMO_HOME' in os.environ:
    tools = os.path.join(os.environ['SUMO_HOME'], 'tools')
    sys.path.append(tools)
else:
    sys.exit("please declare environment variable 'SUMO_HOME'")

from sumolib import checkBinary  # noqa
import traci  # noqa
import sumolib  # noqa


PEOPLE_RESERVED = []
TOTAL_TAXI = 0


# Creating 4 taxis
def createTaxi(time, taxi_num):
    global TOTAL_TAXI
    try:
        fleet = traci.vehicle.getTaxiFleet(0)
    except (traci.exceptions.FatalTraCIError):
        print("No unoccupied taxi-fleet!")
    else:
        print(f"Total taxi:: {TOTAL_TAXI}")
        for i in range(TOTAL_TAXI,  taxi_num + TOTAL_TAXI):
            # declaring the name(unique), route(from demand.route.xml), type of vehicle(declared in demand.route.xml),
            # depart time, and line
            print(f"Add taxiV{i}")
            traci.vehicle.add('taxiV%i' % i, 'route_0',
                            'taxi', depart=time, line='taxi')
        TOTAL_TAXI += taxi_num


# Dispatching taxis to cater to people waiting at a busstop
def emergencyTaxi(busstopID):
    # getting a Id-list of people waiting at the busstop
    peopleWaiting = traci.busstop.getPersonIDs(busstopID)
    print("Person Waiting: \n", peopleWaiting)

    # for people in peopleToReserved:
    #     if people not in PEOPLE_RESERVED:
    #         PEOPLE_RESERVED.append(people)
    try:
        available_fleet = traci.vehicle.getTaxiFleet(0)
    except (traci.exceptions.FatalTraCIError):
        print("No unoccupied taxi-fleet!")

    if len(peopleWaiting) > len(available_fleet):
        print("Insufficient fleet! Remaining: ", len(available_fleet))
        people_to_pickup = peopleWaiting[:len(available_fleet)]
        fleet = available_fleet
    
    elif len(peopleWaiting) < len(available_fleet):
        print("Excess fleet! Remaining: ", len(available_fleet))
        people_to_pickup = peopleWaiting
        fleet = available_fleet[:len(peopleWaiting)]
    
    else:
        print("Balanced fleet! Remaining: ", len(available_fleet))
        people_to_pickup = peopleWaiting
        fleet = available_fleet
    
    print("Fleet match people to reserve: ", len(fleet)==len(people_to_pickup))

    pickup = []
    print("Full reservation list: \n", traci.person.getTaxiReservations(0))
    print("New reservation list: \n", traci.person.getTaxiReservations(1))
    print("People in Already Reserved: \n", PEOPLE_RESERVED)
    curr_reservations = traci.person.getTaxiReservations(0)
    people_to_reserve = []
    # creating a list with the taxi resevations
    for i, p_id in enumerate(people_to_pickup):
        if len(PEOPLE_RESERVED) == 0:
            target_idx = next((i for i, reservation in enumerate(curr_reservations) if p_id in reservation.persons), None)
            pickup.append(traci.person.getTaxiReservations(0)[target_idx].id)
            PEOPLE_RESERVED.append(p_id)
            people_to_reserve.append(p_id)

        else:
            if p_id not in PEOPLE_RESERVED:
                target_idx = next((i for i, reservation in enumerate(curr_reservations) if p_id in reservation.persons), None)
                pickup.append(traci.person.getTaxiReservations(0)[target_idx].id)
                PEOPLE_RESERVED.append(p_id)
                people_to_reserve.append(p_id)

    # if one Taxi should pick up all customers, the list needs to clarify the drop off
    # hence the pickup is extendet by the order of drop offs
    # pickup.extend(pickup)
    print("People to pickup: \n", people_to_reserve)
    print("PickupList:\n", pickup)

    if len(pickup) > 0 and len(people_to_reserve) <= len(fleet):
        # dispatching the unoccupied taxis to pick up their designated customers
        for i in range(len(people_to_reserve)):
            print("Taxifleet-Status")
            print("empty: ", traci.vehicle.getTaxiFleet(0))
            print("pickup: ", traci.vehicle.getTaxiFleet(1))
            print(f"Taxi chosen: {fleet[i]} for Pickup ID: {pickup[i]}")
            traci.vehicle.dispatchTaxi(fleet[i], pickup[i])


# checks if busstop is "overflowing"
def busstopCheck(time):
    # getting all bus stops on the map
    busstops = traci.busstop.getIDList()
    # checking the personcount of waiting people
    for i in busstops:
        if traci.busstop.getPersonCount(i) >= 8:
            try:
                fleet = traci.vehicle.getTaxiFleet(0)
            except (traci.exceptions.FatalTraCIError):
                print("No unoccupied taxi-fleet!")
            
            print(f"Total Taxi Count: {TOTAL_TAXI}")
            if TOTAL_TAXI < traci.busstop.getPersonCount(i):
                print("Create Taxi to meet demand")
                createTaxi(time, traci.busstop.getPersonCount(i) - TOTAL_TAXI)

            # emergencyTaxi(i)

def dispatch_taxi():
    # getting all bus stops on the map
    busstops = traci.busstop.getIDList()
    # checking the personcount of waiting people
    for i in busstops:
        if traci.busstop.getPersonCount(i) > 0:
            emergencyTaxi(i)


def run():
    """execute the TraCI control loop"""
    step = 0
    while traci.simulation.getMinExpectedNumber() > 0:
        traci.simulationStep()
        step += 1
        check = step%10
        dispatch = step%25
        print(f"Current step: {step}")
        if step == 20:
            print("First taxi creation")
            createTaxi(20, 4)

        if check == 0 and dispatch != 0:
            print("Check busstop")
            busstopCheck(step)

        if dispatch == 0 and check != 0:
            print("Dispatch taxi")
            dispatch_taxi()

        if step == 5000:
            break
    traci.close()
    sys.stdout.flush()


def get_options():
    optParser = optparse.OptionParser()
    optParser.add_option("--nogui", action="store_true",
                         default=False, help="run the commandline version of sumo")
    options, args = optParser.parse_args()
    return options


if __name__ == "__main__":
    # first, generate the route file for this simulation
    options = get_options()
    # If you want to run this tutorial please uncomment following lines, that define the sumoBinary
    # and delete the line before traci.start, to use the gui
    if options.nogui:
       sumoBinary = checkBinary('sumo')
    else:
       sumoBinary = checkBinary('sumo-gui')
    # this is the normal way of using traci. sumo is started as a
    # subprocess and then the python script connects and runs
    # sumoBinary = checkBinary('sumo')
    traci.start([sumoBinary, "-c", "data/sumo.sumocfg",
                 "--tripinfo-output", "data/tripinfo.xml"])
    run()
```

**SUMO-version:**

```
Eclipse SUMO sumo Version 1.19.0
 Build features: Windows-10.0.17763 AMD64 MSVC 19.29.30133.0 Release FMI Proj GUI Intl SWIG
 Copyright (C) 2001-2023 German Aerospace Center (DLR) and others; https://sumo.dlr.de
 License EPL-2.0: Eclipse Public License Version 2 <https://eclipse.org/legal/epl-v20.html>
 Use --help to get the list of options.
```

**operating system: Windows 10**

**Python Version: 3.9.13**

**Comment by m-kro** [comment]
OK I cannot reproduce it at the exact same time step but encounter similar warning and error messages. In debug mode the simulation even seems to run into an infinite loop as the time step does not pass....?!

**Comment by apecia** [comment]
> OK I cannot reproduce it at the exact same time step but encounter similar warning and error messages. In debug mode the simulation even seems to run into an infinite loop as the time step does not pass....?!

That's weird.... I think I've set a condition to break at the 5000th steps. I was able to break the loops where I run those simulations so I not quite sure what's the problem here. But I'm glad that you were able to encounter similar warning and error messages because I've been trying to debug these for days.

**Comment by apecia** [comment]
> > OK I cannot reproduce it at the exact same time step but encounter similar warning and error messages. In debug mode the simulation even seems to run into an infinite loop as the time step does not pass....?!
> 
> That's weird.... I think I've set a condition to break at the 5000th steps. I was able to break the loops when I run those simulations (Mine just crashed before it can get into infinite loop) so I not quite sure what's the problem here. But I'm glad that you were able to encounter similar warning and error messages because I've been trying to debug these for days.

On second thought, could it be a CPU or RAM issue? I'm running these simulations on my laptop with 4 cores and 8GB of RAM. Maybe that explains why mine just crashed before it enters an infinite loop?

**Comment by apecia** [comment]
I did more simulations and find out the error message seems to be caused by the method `traci.vehicle.dispatchTaxi()`. The following is my debug messages:

```
Taxi chosen: taxiV0 for Pickup ID: 12
Traceback (most recent call last):
  File "C:\Users\APE7280-1\Documents\Taxi Service\sumo_test\runner.py", line 204, in <module>
    run()
  File "C:\Users\APE7280-1\Documents\Taxi Service\sumo_test\runner.py", line 173, in run
    dispatch_taxi()
  File "C:\Users\APE7280-1\Documents\Taxi Service\sumo_test\runner.py", line 151, in dispatch_taxi
    emergencyTaxi(i)
  File "C:\Users\APE7280-1\Documents\Taxi Service\sumo_test\runner.py", line 123, in emergencyTaxi
    traci.vehicle.dispatchTaxi(fleet[i], pickup[i])
  File "C:\Program Files (x86)\Eclipse\Sumo\tools\traci\_vehicle.py", line 1469, in dispatchTaxi
    self._setCmd(tc.CMD_TAXI_DISPATCH, vehID, "l", reservations)
  File "C:\Program Files (x86)\Eclipse\Sumo\tools\traci\domain.py", line 164, in _setCmd
    self._connection._sendCmd(self._cmdSetID, varID, objectID, format, *values)
  File "C:\Program Files (x86)\Eclipse\Sumo\tools\traci\connection.py", line 231, in _sendCmd
    return self._sendExact()
  File "C:\Program Files (x86)\Eclipse\Sumo\tools\traci\connection.py", line 144, in _sendExact
    raise TraCIException(err, prefix[1], _RESULTS[prefix[2]])
traci.exceptions.TraCIException: Could not interpret reservations for vehicle 'taxiV0' (Reservation id '2' is not known).
```
I passed in a `Reservation id 12` but the error message were getting `Reservation id 2`. I think there is some bugs in the parsing method for the `dispatchTaxi` method.

**Comment by apecia** [comment]
Hi, I'd like to stay informed about the progress of this issue. If there's any way I can contribute to resolving the problem, please let me know.

**Comment by coopwilliams** [comment]
I encountered this error as well, when passing the ID directly to `dispatchTaxi()` as the second argument. I found that if the pending reservations had IDs 6, 9, and 10, then the function would accept 66 and 99, but error receiving IDs 77 or 10.

**What fixed it for me, was passing instead a list of IDs.**
```
traci.vehicle.dispatchTaxi(taxi_id, [reservation.id])
```

Looking at the function signature [here](https://github.com/eclipse-sumo/sumo/blob/ed5c945fa19158b7b3940c46f7aed5465430e3a1/src/libtraci/Vehicle.cpp#L1325), an array is expected.
